lib/commands/test.ts

```diff
@@ -1,30 +1,73 @@
-		this.canExecute = (args: string[]): Promise<boolean> => canExecute({ $platformEnvironmentRequirements, $projectData, $options, platform });
-		this.allowedParameters = [];
-	};
-}
+abstract class TestCommandBase {
+	public allowedParameters: ICommandParameter[] = [];
+	private projectFilesConfig: IProjectFilesConfig;
+	protected abstract $projectData: IProjectData;
+	protected abstract $testExecutionService: ITestExecutionService;
+	protected abstract $analyticsService: IAnalyticsService;
+	protected abstract $options: IOptions;
+	protected abstract $platformEnvironmentRequirements: IPlatformEnvironmentRequirements;
+	protected abstract $errors: IErrors;
+
+	constructor(private platform: string) {
+	}
+
+	async execute(args: string[]): Promise<void> {
+		await this.$testExecutionService.startKarmaServer(this.platform, this.$projectData, this.projectFilesConfig);
+	}
 
-async function canExecute(input: { $platformEnvironmentRequirements: IPlatformEnvironmentRequirements, $projectData: IProjectData, $options: IOptions, platform: string }): Promise<boolean> {
-	const { $platformEnvironmentRequirements, $projectData, $options, platform } = input;
-	const output = await $platformEnvironmentRequirements.checkEnvironmentRequirements({
-		platform,
-		projectDir: $projectData.projectDir,
-		options: $options,
-		notConfiguredEnvOptions: {
-			hideSyncToPreviewAppOption: true,
-			hideCloudBuildOption: true
+	async canExecute(args: string[]): Promise<boolean | ICanExecuteCommandOutput> {
+		this.$projectData.initializeProjectData();
+		this.$analyticsService.setShouldDispose(this.$options.justlaunch || !this.$options.watch);
+		this.projectFilesConfig = helpers.getProjectFilesConfig({ isReleaseBuild: this.$options.release });
+
+		const output = await this.$platformEnvironmentRequirements.checkEnvironmentRequirements({
+			platform: this.platform,
+			projectDir: this.$projectData.projectDir,
+			options: this.$options,
+			notConfiguredEnvOptions: {
+				hideSyncToPreviewAppOption: true,
+				hideCloudBuildOption: true
+			}
+		});
+
+		const canStartKarmaServer = await this.$testExecutionService.canStartKarmaServer(this.$projectData);
+		if (!canStartKarmaServer) {
+			this.$errors.fail({
+				formatStr: "Error: In order to run unit tests, your project must already be configured by running $ tns test init.",
+				suppressCommandHelp: true,
+				errorCode: ErrorCodes.TESTS_INIT_REQUIRED
+			});
 		}
-	});
 
-	return output.canExecute;
+		return output.canExecute && canStartKarmaServer;
+	}
+}
+
+class TestAndroidCommand extends TestCommandBase implements ICommand {
+	constructor(protected $projectData: IProjectData,
+		protected $testExecutionService: ITestExecutionService,
+		protected $analyticsService: IAnalyticsService,
+		protected $options: IOptions,
+		protected $platformEnvironmentRequirements: IPlatformEnvironmentRequirements,
+		protected $errors: IErrors) {
+		super("android");
```

Comment:
Maybe use "protected abstract" in parent class and just set in derived classes.

Reply:
Using constant for the platform will be good improvement.

Reply:
https://github.com/NativeScript/nativescript-cli/issues/4388